public/js/components/specs/app.spec.js

```diff
@@ -0,0 +1,243 @@
+    const { vm } = shallowMountApp();
+
+    vm.dashboard.bgImageRepeat = mockDashboardData.dashboard.bgImageRepeat;
+
+    expect(vm.dashStyle).toEqual({
+      backgroundColor: vm.dashboard.bgColor,
+      backgroundImage: vm.dashboard.bgImageUrl,
+      backgroundRepeat: "repeat"
+    });
+
+    vm.dashboard.bgImageRepeat = false;
+
+    expect(vm.dashStyle).toEqual({
+      backgroundColor: vm.dashboard.bgColor,
+      backgroundImage: vm.dashboard.bgImageUrl,
+      backgroundRepeat: "no-repeat"
+    });
+  });
+
+  test("h1 heading style based upon headingStyle computed method", () => {
+    const { vm } = shallowMountApp();
+
+    expect(vm.headingStyle).toEqual({ color: "#000" });
+  });
+
+  test("don't display dashboard-settings based upon showSettings computed method", () => {
+    const wrapper = shallowMountApp();
+
+    expect(wrapper.vm.showSettings).toEqual(true);
+
+    wrapper.vm.dashboard.editMode = "locked";
+
+    expect(wrapper.vm.showSettings).toEqual(false);
+    expect(wrapper.find({ name: "dashboard-settings" }).exists()).toBe(false);
+  });
+
+  test("compute the appTitle with or without an emoji using the TITLE_EMOJI_REGEX constant", () => {
+    const { vm } = shallowMountApp();
+    const title = TITLE_EMOJI_REGEX.exec(mockDashboardData.dashboard.title);
+
+    title[1] = "\u2700";
+    title[2] = "IoT Dashboard";
+
+    expect(vm.appTitle).toEqual(
+      `<span class="hemoji">${title[1]}</span> ${title[2]}`
+    );
+  });
+
+  test("onSaveSettings method", () => {
+    const { vm } = shallowMountApp();
+
+    expect(vm.dashboard).toEqual(
+      expect.objectContaining({
+        bgColor: expect.any(String),
+        bgImageRepeat: expect.any(Boolean),
+        bgImageUrl: expect.any(String),
+        blockSize: expect.any(Array),
+        editMode: expect.any(String),
+        tiles: expect.any(Array),
+        title: expect.any(String)
+      })
+    );
+
+    vm.onSaveSettings(event => {
+      event = {
+        bgColor: "#fff",
+        bgImageRepeat: true,
+        bgImageUrl: "",
+        title: "\u2700 IoT Dashboard"
+      };
+      expect(vm.saveDashboard).toHaveBeenCalled();
+    });
+
+    Vue.nextTick(() => {
+      expect(vm.dashboard).toEqual(
+        expect.objectContaining({
+          bgColor: expect.any(String),
+          bgImageRepeat: expect.any(Boolean),
+          bgImageUrl: expect.any(String),
+          blockSize: expect.any(Array),
+          editMode: expect.any(String),
+          tiles: expect.any(Array),
+          title: expect.any(String)
+        })
+      );
+    });
+  });
+
+  test("onTileChange method", () => {
+    const { vm } = shallowMountApp();
+
+    expect(vm.dashboard).toEqual(mockDashboardData.dashboard);
+
+    mockDashboardData.dashboard.tiles[0].buttonText = "stop it";
+
+    vm.onTileChange(event => {
+      event.id = mockDashboardData.dashboard.tiles[0].id;
+      expect(vm.saveDashboard).toHaveBeenCalled();
```

Comment:
This will fail - I believe the failure is masked by it being called inside the `vm.onTileChange` function. Notes and fix in the above comment.

Reply:
Updated!